### PR TITLE
raftstore-v2:  support dynamic change apply pool size and store io size

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Ref #14485

What's Changed:

```commit-message
 support dynamic change apply pool size and store io size
```

### Related changes

https://github.com/tikv/yatp/pull/81

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by bufferflies — APPROVED** [review]

File: Cargo.toml [hunk 1]
```diff
@@ -202,6 +202,8 @@ procinfo = { git = "https://github.com/tikv/procinfo-rs", rev = "6599eb9dca74229
 # After the PR to kvproto is merged, remember to comment this out and run `cargo update -p kvproto`.
 # [patch.'https://github.com/pingcap/kvproto']
 # kvproto = { git = "https://github.com/your_github_id/kvproto", branch = "your_branch" }
+[patch.'https://github.com/tikv/yatp.git']
```
Comment (bufferflies) [inline Cargo.toml:205]: why change it？

Comment (SpadeA-Tang) [reply Cargo.toml:205]: Temporary change. Make tikv be able to fetch min_thread_count and max_thread_count from ThreadPool.

File: components/raftstore-v2/src/worker/refresh_config.rs [hunk 2]
```diff
@@ -117,13 +133,88 @@ where
...
+                "to" => size
+            );
+        }
+    }
+
+    /// Resizes the count of background threads in store_writers.
+    fn resize_store_writers(&mut self, size: usize) {
+        // The resizing of store writers will not directly update the local
+        // cached store writers in each poller. Each poller will timely
+        // correct its local cached in its next `poller.begin()` after
```
Comment (LykxSassinator) [inline components/raftstore-v2/src/worker/refresh_config.rs:173]: Here, in this pr, why does `raftstore-v2` not need to update the cached senders in `WriteSenders` in `poller.begin() just like what we do in v1?
https://github.com/tikv/tikv/blob/48b18a66b8a469bea1a5621954b9dd5f078cdb4f/components/raftstore/src/store/fsm/store.rs#L923-L924

Comment (SpadeA-Tang) [reply components/raftstore-v2/src/worker/refresh_config.rs:173]: Good catch. I missed it.

**Review by LykxSassinator — APPROVED** [review]

File: Cargo.toml [hunk 3]
```diff
@@ -202,6 +202,8 @@ procinfo = { git = "https://github.com/tikv/procinfo-rs", rev = "6599eb9dca74229
...
+yatp = { git = "https://github.com/SpadeA-Tang/yatp.git", branch = "thread-limit" }
```
Comment (tonyxuqqi) [inline Cargo.toml:206]: make sure remove it before merge.

File: components/raftstore-v2/src/worker/refresh_config.rs [hunk 4]
```diff
@@ -117,13 +133,88 @@ where
...
+        self.apply_pool.scale_pool_size(size);
+        let (min_thread_count, max_thread_count) = self.apply_pool.thread_count_limit();
+        if size > max_thread_count || size < min_thread_count {
+            warn!(
+                self.logger,
+                "apply pool scale size is out of bound, and the size is clamped";
+                "size" => size,
+                "min_thread_limit" => min_thread_count,
+                "max_thread_count" => max_thread_count,
+            );
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/worker/refresh_config.rs:158]: Should we check the size before actually scale_pool_size is called.  If the size is out of count limit, we just log error and fail the operation.

**Review by tonyxuqqi — APPROVED** [review]

File: tests/integrations/raftstore/test_scale_pool.rs [hunk 5]
```diff
@@ -198,6 +199,72 @@ fn test_decrease_pool() {
...
+        let sim = cluster.sim.rl();
+        let cfg_controller = sim.get_cfg_controller().unwrap();
+        let change = {
+            let mut change = HashMap::new();
+            change.insert("raftstore.apply-pool-size".to_owned(), "4".to_owned());
+            change
+        };
+
+        cfg_controller.update(change).unwrap();
+        std::thread::sleep(std::time::Duration::from_secs(1));
```
Comment (tonyxuqqi) [inline tests/integrations/raftstore/test_scale_pool.rs:231]: How to check the apply pool size is actually 4 after the update.

File: tests/integrations/raftstore/test_scale_pool.rs [hunk 6]
```diff
@@ -198,6 +199,72 @@ fn test_decrease_pool() {
...
+        let sim = cluster.sim.rl();
+        let cfg_controller = sim.get_cfg_controller().unwrap();
+        let change = {
+            let mut change = HashMap::new();
+            change.insert("raftstore.apply-pool-size".to_owned(), "1".to_owned());
+            change
+        };
+
+        cfg_controller.update(change).unwrap();
+        std::thread::sleep(std::time::Duration::from_secs(1));
```
Comment (tonyxuqqi) [inline tests/integrations/raftstore/test_scale_pool.rs:259]: Same here. May be need to check if the apply pool size is actually changed.

File: components/raftstore-v2/src/worker/refresh_config.rs (hunk 4 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/worker/refresh_config.rs:158]: When code reaches here, the actual toml file has been changed.
I think clamp  is fine as there's no use case to exceed the max thread count which is a large value (max of 4 and cpu core numbers).

File: Cargo.toml (hunk 3 above)
Comment (SpadeA-Tang) [reply Cargo.toml:206]: sure.

File: tests/integrations/raftstore/test_scale_pool.rs (hunk 5 above)
Comment (SpadeA-Tang) [reply tests/integrations/raftstore/test_scale_pool.rs:231]: It's hard to directly verify the pool size as the yatp generally creates more threads than needed.
Here I block one pool thread and make sure apply will be blocked. Then I change the pool size and after that, the following apply will not be blocked.

Comment (tonyxuqqi) [reply tests/integrations/raftstore/test_scale_pool.rs:231]: How about use fail point in resize_apply_pool and let the code pass the thread count to that fail point. 
Then we can verify if that thread count is correct.

**Pull request merged**